test/actual-uploads.test.js

```diff
@@ -0,0 +1,393 @@
+      signer,
+      hash,
+      data,
+      bucket,
+    }) {
+      const url = signer.sign({
+        bucket,
+        key: `testing/test-file-${Date.now()}.json`,
+        checksum: hash,
+        expires: 1000,
+      })
+
+      const rsp = await fetch(url.toString(), {
+        method: 'PUT',
+        body: JSON.stringify(data),
+        headers: {
+          'x-amz-checksum-sha256': hash + 'ss',
+        },
+      })
+      const out = await rsp.text()
+      assert.ok(out.includes('SignatureDoesNotMatch'))
+    })
+
+    it('should sign and fail upload because expired', async function ({
+      signer,
+      hash,
+      data,
+      bucket,
+    }) {
+      const url = signer.sign({
+        bucket,
+        key: `testing/test-file-${Date.now()}.json`,
+        checksum: hash,
+        expires: 1,
+      })
+
+      await sleep(1000)
+      const rsp = await fetch(url.toString(), {
+        method: 'PUT',
+        body: JSON.stringify(data),
+        headers: {
+          'x-amz-checksum-sha256': hash,
+        },
+      })
+      const out = await rsp.text()
+      assert.ok(out.includes('Request has expired'))
+    })
+
+    it('should sign and succeed upload with content length given', async function ({
+      signer,
+      hash,
+      data,
+      bucket,
+    }) {
+      const content = JSON.stringify(data)
+      const contentLength = Buffer.from(content).byteLength
+      const url = signer.sign({
+        bucket,
+        key: `testing/test-file-${Date.now()}.json`,
+        checksum: hash,
+        expires: 1000,
+        contentLength,
+      })
+
+      await sleep(500)
+      const rsp = await fetch(url.toString(), {
+        method: 'PUT',
+        body: content,
+        headers: {
+          'content-length': contentLength,
+          'x-amz-checksum-sha256': hash,
+        },
+      })
+      const out = await rsp.text()
+      console.log(out)
+
+      assert.ok(rsp.ok)
+      //       assert.ok(out.includes('Request has expired'))
+    })
+
+    it('should sign and upload when content length does match with "bad fetch".', async function ({
+      signer,
+      hash,
+      data,
+      bucket,
+    }) {
+      const content = JSON.stringify(data)
+      const contentLength = Buffer.from(content).byteLength
+      const url = signer.sign({
+        bucket,
+        key: `testing/test-file-${Date.now()}-invalid-size.json`,
+        checksum: hash,
+        expires: 1000,
+        contentLength,
+      })
+
+      const rsp = await badFetch(url, {
+        method: 'PUT',
+        body: content,
+        contentEncoding: 'base64',
+        contentType: 'application/json',
+        headers: {
+          'content-length': contentLength, //actual size is 15
+          'x-amz-checksum-sha256': hash,
+        },
+      })
+      assert.ok(rsp.statusCode == 200)
+    })
+
+    it.fails(
+      'should sign and fail upload when content length is greater than actual content "bad fetch".',
```

Comment:
The suffixing of `bad fetch` and `manual writing with socket` might read nicely if split out into separate `describe` blocks like 

```
describe('fetch', function () {
  it(''should sign and fail upload when content length is greater than actual content', ...
  it('should sign and upload when content length does match', ...
}
describe('http.request', function () {
  it(''should sign and fail upload when content length is greater than actual content', ...
  it('should sign and upload when content length does match', ...
}
describe('net.Socket', function () {
  it(''should sign and fail upload when content length is greater than actual content', ...
  it('should sign and upload when content length does match', ...
}
```